src/iterative.F90

```diff
@@ -1371,324 +1108,170 @@ subroutine Gr_blk2csr(P,nbl,indblk,A)
-      ! C = C + A*B
-      if (j.NE.ref .AND. ABS(frm(j)-frm(ref)).GT.EPS) THEN
+    if (allocated(negf%inter)) then
+      call negf%inter%get_sigma_n(Sigma_n, negf%ie)
+    end if
 
+    do j=1,ncont
+      frmdiff = frm(j) - frm(ref)
+      if (j.NE.ref .AND. ABS(frmdiff).GT.EPS) THEN
         cb=struct%cblk(j) ! block corresponding to contact j
-
         call zspectral(SelfEneR(j),SelfEneR(j),0,Gam)
+        Sigma_n(cb,cb)%val = Sigma_n(cb,cb)%val + frmdiff*Gam%val
+        call destroy(Gam)
+      endif
+    end do
 
-        frmdiff = cmplx(frm(j)-frm(ref),0.0_dp,dp)
-        ! Computation of Gl(1,1) = Gr(1,cb) Gam(cb) Ga(cb,1)
-        if (allocated(Gr(1,cb)%val)) then
-          call prealloc_mult(Gr(1,cb),Gam,work1)
-          call zdagger(Gr(1,cb),Ga)
-          call prealloc_mult(work1,Ga,frmdiff,Gn(1,1))
-          call destroy(work1, Ga)
-        else
-          Gn(1,1)%val=(0.0_dp,0.0_dp)
-        endif
+    call calculate_sigma_n()
 
-        ! Computation of all tridiagonal blocks
-        do i=2,nbl
+    call zdagger(Gr(1,1), Ga)
+    call prealloc_mult(Sigma_n(1,1), Ga, work1)
+    call prealloc_mult(Gr(1,1), work1, Gn(1,1))    
+    call destroy(Ga, work1)
 
-          ! Computation of Gl(i,j) = Gr(i,cb) Gam(cb) Ga(cb,j)
-          ! Both Gr(i,cb) and Gr(j,cb) must be non-zero
-          if (Gr(i-1,cb)%nrow.gt.0 .and. Gr(i,cb)%nrow.gt.0) then
-            call prealloc_mult(Gr(i-1,cb),Gam,work1)
-            call zdagger(Gr(i,cb),Ga)
-            call prealloc_mult(work1,Ga,frmdiff,Gn(i-1,i))
-            call destroy(work1)
+    do i = 1, nbl-1
+        !Gn(i+1,i) = gsmr(i+1)*[Sigma(i+1,i)Ga(i,i) + Sigma(i+1,i+1)Ga(i+1,i) - Tr(i+1,i)Gn(i,i)]
+        call zdagger(Gr(i,i), Ga)
+        call prealloc_mult(Sigma_n(i+1,i), Ga, work1)
+        call destroy(Ga)
 
-            call prealloc_mult(Gr(i,cb),Gam,work1)
-            call prealloc_mult(work1,Ga,frmdiff,Gn(i,i))
+        call zdagger(Gr(i,i+1), Ga)
+        call prealloc_mult(Sigma_n(i+1,i+1), Ga, work2)
+        call destroy(Ga)
 
-            call destroy(work1, Ga)
+        call prealloc_mult(ESH(i+1,i), Gn(i,i), work3)
+        call create(factors, work1%nrow, work1%ncol)
+        factors%val = work1%val + work2%val - work3%val
 
-            call prealloc_mult(Gr(i,cb),Gam,work1)
-            call zdagger(Gr(i-1,cb),Ga)
-            call prealloc_mult(work1,Ga,frmdiff,Gn(i,i-1))
+        call prealloc_mult(gsmr(i+1), factors, Gn(i+1,i))
+        call destroy(factors, work1, work2, work3)
 
-            call destroy(work1, Ga)
-          else
-            Gn(i  ,i)%val=(0.0_dp,0.0_dp)
-            Gn(i-1,i)%val=(0.0_dp,0.0_dp)
-            Gn(i,i-1)%val=(0.0_dp,0.0_dp)
-          endif
+        !Gn(i,i+1) = [Gr(i,i)Sigma(i,i+1) + Gr(i,i+1)Sigma(i+1,i+1) - Gn(i,i)Ta(i,i+1)] * gsma(i+1)
+        !Gn(i,i+1) = Gn(i+1,i)^dag
+        call zdagger(Gn(i+1,i), work1)
+        Gn(i,i+1)%val = work1%val
+        call destroy(work1)
 
```

Comment:
can clean up these commented lines? Or if they're here as reminder of something, maybe leave a comment why they are here